File: prince-of-versions/src/main/java/co/infinum/princeofversions/InteractorImpl.java
package co.infinum.princeofversions;

import android.util.Log;

final class InteractorImpl implements Interactor {

    private ConfigurationParser configurationParser;

    /**
     * Constructs interactor using provided {@link ConfigurationParser}.
     *
     * @param configurationParser object which will be used for parsing update resource.
     */
    InteractorImpl(ConfigurationParser configurationParser) {
        this.configurationParser = configurationParser;
    }

    @Override
    public CheckResult check(final Loader loader, final ApplicationConfiguration appConfig) throws Throwable {
        String content = loader.load();
        PrinceOfVersionsConfig config = configurationParser.parse(content);
        Log.d("META:", config.getMetadata().toString());

        int currentVersion = appConfig.version();

        Integer mandatoryConfigVersion = config.getMandatoryVersion();
        Integer optionalConfigVersion = config.getOptionalVersion();

        if (mandatoryConfigVersion == null && optionalConfigVersion == null) {
            // neither mandatory nor optional version is provided
            throw new IllegalStateException("Both mandatory and optional version are null.");
        }

        if (mandatoryConfigVersion != null) {
            int mandatoryVersion = mandatoryConfigVersion;

            if (currentVersion < mandatoryVersion) {
                // if mandatory update exists - notify mandatory update
                // if there is also optional update available check if its version is greater than mandatory
                // in that case notify mandatory update with optional version, otherwise notify mandatory update with mandatory version
                if (optionalConfigVersion != null) {
                    int optionalVersion = optionalConfigVersion;
                    if (optionalVersion > mandatoryVersion) {
                        // optional update also exists and has greater version than mandatory
                        return CheckResult.mandatoryUpdate(optionalVersion, config.getMetadata());
                    }
                }
                // if there is no optional update or it isn't greater than mandatory - notify mandatory version
                return CheckResult.mandatoryUpdate(mandatoryVersion, config.getMetadata());
            }
        }

        // if there is no mandatory update check for optional
        if (optionalConfigVersion != null) {
            int optionalVersion = optionalConfigVersion;
            if (currentVersion < optionalVersion) {
                return CheckResult.optionalUpdate(optionalVersion, config.getOptionalNotificationType(),
                    config.getMetadata());
            }
        }

        return CheckResult.noUpdate(currentVersion, config.getMetadata());
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/InteractorImpl.kt
package co.infinum.princeofversions

/**
 * The concrete implementation of the [Interactor] that orchestrates the update check.
 *
 * @param configurationParser The parser used to interpret the update configuration resource.
 */
internal class InteractorImpl(
    private val configurationParser: ConfigurationParser
) : Interactor {

    @Throws(Throwable::class)
    override fun check(loader: Loader, appConfig: ApplicationConfiguration): CheckResult {
        val content = loader.load()
        val config = configurationParser.parse(content)
        val currentVersion = appConfig.version

        val updateInfo = UpdateInfo(
            requiredVersion = config.mandatoryVersion,
            lastVersionAvailable = config.optionalVersion,
            requirements = config.requirements,
            installedVersion = currentVersion,
            notificationFrequency = config.optionalNotificationType
        )

        val mandatoryVersion = config.mandatoryVersion
        val optionalVersion = config.optionalVersion

        // Check for mandatory update first
        if (mandatoryVersion != null && currentVersion < mandatoryVersion) {
            // If a mandatory update is available, it takes precedence.
            // The notified version will be the greater of the optional and mandatory versions.
            val versionToNotify = optionalVersion?.takeIf { it > mandatoryVersion } ?: mandatoryVersion
            return CheckResult.mandatoryUpdate(versionToNotify, config.metadata, updateInfo)
        }

        // If no mandatory update, check for an optional update
        if (optionalVersion != null && currentVersion < optionalVersion) {
            return CheckResult.optionalUpdate(optionalVersion, config.optionalNotificationType, config.metadata, updateInfo)
        }

        // If neither mandatory nor optional update is available, or if no versions are defined, return no update.
        check(!(mandatoryVersion == null && optionalVersion == null)) { "Both mandatory and optional version are null." }

        return CheckResult.noUpdate(currentVersion, config.metadata, updateInfo)
    }
}
